Update emails and timelines for subscriptions nearing their end dates

nospame opened the pull request:
## Product Description
"Renewal Reminder" email:
<img width="1081" height="342" alt="image" src="https://github.com/user-attachments/assets/791e539e-6a45-4088-bf72-68c004c0bf7f" />

"Subscription ending" email:
<img width="668" height="203" alt="image" src="https://github.com/user-attachments/assets/3cdba145-503c-42f4-9692-1ae3a9ded396" />

Updated text of "Dimagi Contact" email:
<img width="1115" height="258" alt="image" src="https://github.com/user-attachments/assets/640e389f-c552-4fdb-b229-a6a34af76f2d" />



## Technical Summary
[Ticket](https://dimagi.atlassian.net/browse/SAAS-18409), [Desired Alert Text](https://docs.google.com/spreadsheets/d/1YFoixutlTSXQXssae5ZfPs2gfPN96hRdW8QHJDV3Uu4/edit?usp=sharing)
In essence, this adds 2 new types of emails and updates the content of another. Along with this, I made an effort at cleaning up the code to send these emails to be a bit DRYer and pull it off the `Subscription` model.

There's a good number of changes here which I have attempted to keep in manageable commits. Review by commit highly suggested.

## Feature Flag
SHOW_AUTO_RENEWAL for most of these changes.

## Safety Assurance

### Safety story
Tested locally and on staging with the additions to the Trigger Reminder Email form to see that emails are sent when expected with the desired content.

### Automated test coverage
I struggled a bit with how to test here, since most of our email sending is a good bit simpler than this. I opted for basic automated tests on the conditions for sending particular emails out of `tasks.py` (`TestSubscriptionReminderEmail`), as well as enabling manual testing of emails through the UI-based Trigger Reminder Email accounting testing tool. Also added `TestGetReminderEmailContacts` and `TestEndingReminderContext` to see that the expected contacts and context are used for subscription ending emails.

### QA Plan
This will be included in QA for the Auto Renewals feature before the feature flag is removed.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Refactor to _filter_subscriptions_ending_in_n_days

nospame committed:
Add dummy functions to send new subscription reminder emails

nospame committed:
Refactor into _try_send_subscription_email and add stub methods for new emails

nospame committed:
Remove unused trial ending reminder

We already filter out these subscriptions with is_trial=False in the send_subscription_reminder_emails task

nospame committed:
Remove checks against unreasonable conditions

We already filter against these in the respective _reminder_email tasks

nospame committed:
Filter out Paused subscriptions when fetching initially

nospame committed:
Add new subscription ending email templates

nospame committed:
Add methods to send new reminder emails with shared context

nospame committed:
Send new subscription ending emails collectively to contacts

nospame committed:
Rename function -> _filter_subscriptions_for_reminder_emails

nospame committed:
Move all Subscription emails to emails.py

nospame committed:
Nix separate customer billing account reminder for Dimagi contacts

nospame committed:
Update content of Dimagi subscription ending email

nospame committed:
Add tests to assert criteria for sending emails

nospame committed:
Allow triggering other Subscription emails from Trigger Reminder Email form

nospame committed:
Tweak text formatting in emails

nospame committed:
Remove user email addresses from logging

Only one of these was flagged by CodeQL, but they're essentially equivalent

nospame committed:
Hyperlink 'here' in subscription ending html email

nospame committed:
Add TestGetReminderEmailContacts

nospame committed:
Reject any falsy value for account.dimagi_contact

nospame committed:
Clean up subject in _ending_reminder_context

nospame committed:
Add TestEndingReminderContext

mjriley commented on corehq/apps/accounting/models.py:
```diff
@@ -1684,10 +1684,45 @@ def transfer_credits(self, subscription=None):
+
+    def _send_ending_reminder_email(self, template, template_plaintext):
+        today = datetime.date.today()
+        num_days_left = (self.date_end - today).days
```
Nit: It looks like we only use this when logging, so it probably isn't severe, but...the day when this reminder actually gets sent out, vs the day that we fetched this subscription may not align. I.e. 1 minute before midnight, we might fetch all subscriptions set to expire in 10 days. If the iteration took long enough, by the time we process a subscription, it might be the next day, and then we'd log that we sent a "9 day subscription reminder", which isn't one of our defined periods. Could we pass the actual date period instead?

mjriley commented on corehq/apps/accounting/models.py:
```diff
@@ -1703,26 +1703,54 @@ def _send_ending_reminder_email(self, template, template_plaintext):
 
         email_html = render_to_string(template, context)
         email_plaintext = render_to_string(template_plaintext, context)
-        bcc = [settings.ACCOUNTS_EMAIL] if not self.is_trial else []
```
I don't see this condition accounted for in the new code. Is this no longer necessary?

nospame replied:
That's right. Ops doesn't see value in being BCC'd on all of these, probably especially as our self-service user base grows.

mjriley commented on corehq/apps/accounting/tests/test_tasks.py:
```diff
@@ -51,3 +56,111 @@ def test_calculate_form_submitting_mobile_workers_in_all_domains(self):
+        toggles.SHOW_AUTO_RENEWAL.set(self.domain.name, True, namespace=toggles.NAMESPACE_DOMAIN)
+
+    def test_sends_renewal_reminder_email(self):
+        self.subscription.date_end = datetime.date.today() + datetime.timedelta(days=60)
```
I'd feel more comfortable if we used freezegun to specify a hard date. While unlikely, I could see these tests returning failures if run on day boundaries. The subscription could be saved with one date, while `send_renewal_reminder_emails` could be called with another. Using freezegun would eliminate that possibility

mjriley replied:
If this is meant to basically be set in the background (i.e. the tests don't care about the specific date), it probably is cleaner to use the [class decorator](https://time-machine.readthedocs.io/en/latest/usage.html#class-decorator)

nospame replied:
I did this initially, but found when I tried to set `tick=False` it would cause the tests to time out and fail. Not happening any more though, must have been something wonky in my local environment. https://github.com/dimagi/commcare-hq/pull/37072/commits/8b8eafd5e4e81943f2cba96ed2b4c817310ac327

mjriley commented on corehq/apps/accounting/tests/test_emails.py:
```diff
@@ -0,0 +1,78 @@
+
+        project_admins = ['carly@example.com', 'dave@example.com']
+        for email in project_admins:
+            user = WebUser.create(self.domain.name, email, 'test123', None, None, email)
```
This looks like we aren't cleaning up the users created here. This opens the door for these tests to create unpredictable behavior in other tests, if those tests happen to use these same users. Can we add a cleanup that removes them?

mjriley reviewed (commented):
Very nice refactoring on a lot of these areas. Mainly looking for clarification in a few areas. However, I do think the tests need to clean up their users.

nospame commented on corehq/apps/accounting/tasks.py:
```diff
@@ -449,20 +437,33 @@ def send_subscription_reminder_emails(num_days):
-                            .exclude(account__dimagi_contact=''))
+    ending_subscriptions = _filter_subscriptions_ending_in_n_days(
+        num_days, is_active=True
+    ).exclude(account__dimagi_contact='')
```
I did rename in a later commit to try to make it clearer this is returning relevant subscriptions for reminder emails: https://github.com/dimagi/commcare-hq/pull/37072/commits/7e2d343aeb7aabf016b7fb1bcbe0364e531b8f20 In terms of the dimagi contact, this function is sending emails _to dimagi contacts_ (where the others are sending emails directly to customers, irrespective of whether they have a dimagi contact) so the particular email is only relevant for subscriptions that do have a dimagi contact - so I think it is reasonable to have them specified in different places.

nospame committed:
Update tests

- Use time_machine.travel
- Use flag_enabled instead of setting toggle in setup

nospame committed:
Clean up code based on PR feedback

- Don't use single letter variable names
- Account for dimagi_contact being an empty string

nospame committed:
Don't recalculate today(); pass in days_left to use consistently

nospame committed:
Clean up _ending_reminder_context a little

No need to get plan_name from the user_facing_description when it is stored on the SoftwarePlan

nospame committed:
Adjust dimagi contact ending reminder

- Only send to service_type Implementation subscriptions
- Send more frequently
- Rename function for clarity

nospame commented on corehq/apps/accounting/models.py:
```diff
@@ -1797,15 +1792,6 @@ def ending_reminder_context(self):
-            raise SubscriptionReminderError(
-                "This subscription has no Dimagi contact."
-            )
-
```
I renamed the function here: https://github.com/dimagi/commcare-hq/pull/37072/commits/28b23d69387a4e3645b67ce7cb0c64335108fd47 and also adjusted how often and for which subscriptions it gets sent, based on feedback from Ops.

mjriley reviewed (approved):
Looks good. The comments I left are non-blocking.

nospame committed:
Use time_machine.travel class decorator

nospame commented on corehq/apps/accounting/emails.py:
```diff
@@ -145,10 +145,9 @@ def _send_subscription_ending_reminder_email(subscription, template, template_pl
     )
     log_accounting_info(
         "Sent %(days_left)s-day subscription reminder "
-        "email for %(domain)s to %(email)s." % {
```
After a little reading, my preference would be for not logging email addresses here. I think it's also less relevant now than it was before, because it's sending one larger email rather than individual emails to (potentially several) different contacts.

And actually, I think this might be why the alert was raised here and not previously - the billing contact emails in the BillingContactInfo don't have to belong to user accounts, they could belong to people who have never even used CommCare in their lives and just process a wire payment when an invoice comes in, so even better not to log those in particular.

nospame committed:
Re-add check against empty dimagi_contact
